From deb6694a8eedd9b20f8c245cce9d57614cbc4d32 Mon Sep 17 00:00:00 2001
From: ruslanbay <67730802+ruslanbay@users.noreply.github.com>
Date: Tue, 10 Feb 2026 20:04:03 +0100
Subject: [PATCH 37/54] media: v4l2-subdev: Fix invalid array indexing on routing.routes

The 'routes' member in struct v4l2_subdev_routing is a __u64, not a pointer type.
Direct array indexing is not possible. Cast the address to a local struct
v4l2_subdev_route pointer to correctly access the routing table and check
for active routes.
---
 drivers/media/pci/intel/ipu-isys-csi2.c   |  3 ++-
 drivers/media/pci/intel/ipu-isys-subdev.c | 17 ++++++++++-------
 drivers/media/pci/intel/ipu-isys-video.c  |  6 +++---
 3 files changed, 15 insertions(+), 11 deletions(-)

diff --git a/drivers/media/pci/intel/ipu-isys-csi2.c b/drivers/media/pci/intel/ipu-isys-csi2.c
index 54768b288b44..974becb8247f 100644
--- a/drivers/media/pci/intel/ipu-isys-csi2.c
+++ b/drivers/media/pci/intel/ipu-isys-csi2.c
@@ -424,7 +424,8 @@ static int csi2_link_validate(struct media_link *link)
 	}
 
 	for (i = 0; i < routing.num_routes; i++) {
-		if (routing.routes[i].flags & V4L2_SUBDEV_ROUTE_FL_ACTIVE)
+		struct v4l2_subdev_route *route = &r[i];
+		if (route->flags & V4L2_SUBDEV_ROUTE_FL_ACTIVE)
 			active++;
 	}
 
diff --git a/drivers/media/pci/intel/ipu-isys-subdev.c b/drivers/media/pci/intel/ipu-isys-subdev.c
index f294c0ef951f..19c47d2b606f 100644
--- a/drivers/media/pci/intel/ipu-isys-subdev.c
+++ b/drivers/media/pci/intel/ipu-isys-subdev.c
@@ -566,35 +566,38 @@ int ipu_isys_subdev_get_routing(struct v4l2_subdev *sd,
 				struct v4l2_subdev_routing *route)
 {
 	struct ipu_isys_subdev *asd = to_ipu_isys_subdev(sd);
+	struct v4l2_subdev_route *routes = (struct v4l2_subdev_route *)(uintptr_t)route->routes;
 	int i, j;
 
 	for (i = 0, j = 0; i < min(asd->nstreams, route->num_routes); ++i) {
-		route->routes[j].sink_pad = asd->route[i].sink;
+		routes[j].sink_pad = asd->route[i].sink;
+		
 		if (sd->entity.pads[asd->route[i].sink].flags &
 		    MEDIA_PAD_FL_MULTIPLEX) {
 			int source_pad = asd->route[i].source - asd->nsinks;
 
-			route->routes[j].sink_stream =
+			routes[j].sink_stream =
 			    asd->stream[asd->route[i].sink].
 			    stream_id[source_pad];
 		} else {
-			route->routes[j].sink_stream =
+			routes[j].sink_stream =
 			    asd->stream[asd->route[i].sink].stream_id[0];
 		}
 
-		route->routes[j].source_pad = asd->route[i].source;
+		routes[j].source_pad = asd->route[i].source;
+		
 		if (sd->entity.pads[asd->route[i].source].flags &
 		    MEDIA_PAD_FL_MULTIPLEX) {
-			route->routes[j].source_stream =
+			routes[j].source_stream =
 			    asd->stream[asd->route[i].source].stream_id[asd->
 									route
 									[i].
 									sink];
 		} else {
-			route->routes[j].source_stream =
+			routes[j].source_stream =
 			    asd->stream[asd->route[i].source].stream_id[0];
 		}
-		route->routes[j++].flags = asd->route[i].flags;
+		routes[j++].flags = asd->route[i].flags;
 	}
 
 	route->num_routes = j;
diff --git a/drivers/media/pci/intel/ipu-isys-video.c b/drivers/media/pci/intel/ipu-isys-video.c
index 8b94556b5275..23976d4d83f0 100644
--- a/drivers/media/pci/intel/ipu-isys-video.c
+++ b/drivers/media/pci/intel/ipu-isys-video.c
@@ -770,11 +770,11 @@ static int link_validate(struct media_link *link)
 		goto err_subdev;
 
 	for (i = 0; i < routing.num_routes; i++) {
-		if (!(routing.routes[i].flags & V4L2_SUBDEV_ROUTE_FL_ACTIVE))
+		if (!(r[i].flags & V4L2_SUBDEV_ROUTE_FL_ACTIVE))
 			continue;
 
-		if (routing.routes[i].source_pad == link->source->index)
-			ip->stream_id = routing.routes[i].sink_stream;
+		if (r[i].source_pad == link->source->index)
+			ip->stream_id = r[i].sink_stream;
 
 		active++;
 	}
-- 
2.51.0

